pkg/webhook/policy_test.go

```diff
@@ -703,3 +706,140 @@ func TestValidateConfigResource(t *testing.T) {
 		})
 	}
 }
+
+// Validate the pending requests counter is incremented for
+// requests as they wait for their turn to use opa.
+func Test_PendingValidationRequests(t *testing.T) {
+	opa := fakeOpa{
+		reviewFn: func(ctx context.Context) {
+			<-ctx.Done()
+		},
+	}
+	cfg := &v1alpha1.Config{
+		Spec: v1alpha1.ConfigSpec{
+			Validation: v1alpha1.Validation{
+				Traces: []v1alpha1.Trace{},
+			},
+		},
+	}
+	review := atypes.Request{
+		AdmissionRequest: admissionv1beta1.AdmissionRequest{
+			Kind: metav1.GroupVersionKind{
+				Group:   "",
+				Version: "v1",
+				Kind:    "Pod",
+			},
+			Object: runtime.RawExtension{
+				Raw: []byte(
+					`{"apiVersion": "v1", "kind": "Pod", "metadata": {"name": "acbd","namespace": "ns1"}}`),
+			},
+			Namespace: "ns1",
+		},
+	}
+
+	const maxThreads = 5
+	sem := make(chan struct{}, maxThreads)
+
+	r, err := newStatsReporter()
+	if err != nil {
+		t.Fatalf("creating stats reporter: %v", err)
+	}
+	handler := validationHandler{opa: opa,
+		webhookHandler: webhookHandler{
+			injectedConfig: cfg,
+			client:         &nsGetter{},
+			reader:         &nsGetter{},
+			reporter:       r,
+		},
+		semaphore: sem,
+	}
+
+	firstCtx, firstCancel := context.WithCancel(context.Background())
+	defer firstCancel()
+	var firstBatch sync.WaitGroup
+	firstBatch.Add(maxThreads)
+	for i := 0; i < maxThreads; i++ {
+		go func(wg *sync.WaitGroup) {
+			wg.Done() // Unblocks the initial Wait().
+			_, _ = handler.reviewRequest(firstCtx, review)
+			wg.Done() // Unblocks the final Wait() at test exit.
+		}(&firstBatch)
+	}
+
+	// Wait for the initial goroutines to acquire the block in opa.Review().
+	firstBatch.Wait()
+	firstBatch.Add(maxThreads) // Used for a second Wait() at test exit.
+
+	// All resources should be in use, but no requests should be pending.
+	if pending := handler.pendingValidationRequests(); pending != 0 {
+		t.Errorf("unexpected pending count, got: %d, expected: %d", pending, 0)
+	}
+
+	// Queue requests that should all block and be pending
+	secondCtx, secondCancel := context.WithCancel(context.Background())
+	defer secondCancel()
+	var secondBatch sync.WaitGroup
+	const expectPending = 5
+	secondBatch.Add(expectPending)
+	for i := 0; i < expectPending; i++ {
+		go func(wg *sync.WaitGroup) {
+			defer wg.Done()
+			_, _ = handler.reviewRequest(secondCtx, review)
+		}(&secondBatch)
+	}
+
+	// Eventually, all the above routines should be blocked trying to acquire the validation lock
+	for i := 0; i < 20; i++ {
```

Comment:
Could this be achieved by calling `secondBatch.Wait()`? I suppose there is a race condition as the wg finishes before trying to acquire the lock.

Then again, that same race condition exists for the pending == 0 test above, with the added difficulty of the initial condition equaling the desired final state.

This all LGTM, though polling/timeouts on unit tests make me wary of flakiness.

Reply:
secondBatch.Wait() returns after those goroutines successfully acquire the lock (it's a `defer wg.Done()`), whereas what we're checking for here is that they're blocked waiting for the lock. I don't know how to synchronize on that directly (save injecting some fake lock interface).

Any other ideas? We could also remove the timeout (endless loop) or increase the timeout to address your concerns.

Reply:
I don't think I have a great idea that doesn't involve piping some kind of "done" signal into the lock acquisition code itself, which probably adds more complexity than value.